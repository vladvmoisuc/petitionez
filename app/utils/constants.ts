export const DATE_FORMATS = {
  STANDARD: "D/MM/YYYY",
};

export const ROUTES = {
  MANAGEMENT: "/management",
  CAMPAIGNS: "/campaigns",
  LOGIN: "/login",
};

export const EMAIL_TYPES = {
  NEW: "new",
  OLD: "old",
};

export const LOGIN_STATES = {
  ...EMAIL_TYPES,
  DEFAULT: "default",
  ERROR: "error",
};

export const INTENTS = {
  VALIDATE: "validate",
  GENERATE: "generate",
  DELETE: {
    CAMPAIGN: "delete-campaign",
  },
  UPDATE: {
    CAMPAIGN: "update-campaign",
  },
  CREATE: {
    ENTITY: "create-entity",
    CAMPAIGN: "create-campaign",
  },
};

export const SLUGS = {
  CREATE: "create",
};

export const LINKS = {
  POSTIMAGES: {
    URL: "https://postimages.org/",
    NAME: "postimages.org",
  },
};
