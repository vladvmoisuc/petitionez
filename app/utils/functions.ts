import ro from "~/content/ro.json";
import errors from "~/content/errors.json";
import logs from "~/content/logs.json";

import type { DynamicObject } from "~/utils/types";

export function waitFor(time: number = 500) {
  return new Promise<void>((resolve) =>
    setTimeout(() => {
      resolve();
    }, time)
  );
}

export function replace(data: { [key: string]: string }, content: string) {
  return Object.keys(data).reduce(
    (text, key) => text.replace(new RegExp(`{{${key}}}`, "g"), data[key]),
    content
  );
}

function extractContent(
  source: DynamicObject,
  key: string,
  values?: DynamicObject
): string {
  const content = source[key as keyof typeof source] ?? "";

  return values ? replace(values, content) : content;
}

export function getText(key: string, values?: DynamicObject): string {
  return extractContent(ro, key, values);
}

export function getError(key: string, values?: DynamicObject): string {
  return extractContent(errors, key, values);
}

export function getLog(key: string, values?: DynamicObject): string {
  return extractContent(logs, key, values);
}

export async function getFormData(
  request: Request
): Promise<{ keys: string[]; values: DynamicObject }> {
  const data = await request.formData();
  const keys = [...data.keys()];

  const values = keys.reduce((values, key) => {
    return {
      ...values,
      [key]: data.get(key),
    };
  }, {});

  return {
    keys,
    values,
  };
}
