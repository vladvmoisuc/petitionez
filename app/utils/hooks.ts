import { useState, useEffect, useCallback } from "react";

import type { Transition } from "@remix-run/react/dist/transition";

export const useLocalStorage = <T>(
  key: string
): [T | undefined, typeof setLocalStorage] => {
  const [state, setState] = useState();

  const isClientSide = typeof window !== "undefined";

  useEffect(() => {
    const value = window.localStorage.getItem(key);

    if (isClientSide && value) {
      setState(typeof value === "string" ? JSON.parse(value) : value);
    }
  }, [isClientSide, key]);

  const setLocalStorage = useCallback(
    (data: any) => {
      const value = typeof data !== "string" ? JSON.stringify(data) : data;

      if (isClientSide) {
        window.localStorage.setItem(key, value);

        setState({ ...data });
      }
    },
    [isClientSide, key]
  );

  return [state, setLocalStorage];
};

export const useAlert = (transition: Transition, time: number = 1000) => {
  const [alert, setAlert] = useState(false);

  useEffect(() => {
    if (transition.state === "loading" && transition.type === "actionReload") {
      setAlert(true);

      setTimeout(() => {
        setAlert(false);
      }, time);
    }
  }, [transition.state, transition.type, time]);

  return alert;
};
