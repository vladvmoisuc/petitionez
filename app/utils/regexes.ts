export default {
  EMAIL: /^\S+@\S+\.\S+$/,
  URL: /^(http:\/\/|https:\/\/).+\..+/,
  START_LETTER: /^[a-z]/g,
  DASHES: /-/g,
  SPACES: /\s/g,
};
