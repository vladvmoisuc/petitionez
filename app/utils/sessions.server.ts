import { createCookieSessionStorage } from "@remix-run/node";

const HOUR = 1 * 60 * 60; // second -> minute -> hour

const { getSession, commitSession, destroySession } =
  createCookieSessionStorage({
    cookie: {
      name: "__session",
      httpOnly: true,
      maxAge: HOUR,
      path: "/",
      sameSite: "lax",
      secrets: [process.env.COOKIES_SECRET_KEY as string],
      secure: true,
    },
  });

async function getLoggedUser(request: Request) {
  const session = await getSession(request.headers.get("Cookie"));

  const loggedUserEmail: string = session.get("email") ?? "";
  const loggedUserId: string = session.get("id") ?? "";
  const isAdmin: boolean = session.get("admin") ?? false;

  return { loggedUserEmail, loggedUserId, isAdmin };
}

export { getSession, commitSession, destroySession, getLoggedUser };
