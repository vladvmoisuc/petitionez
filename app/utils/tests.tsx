import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import type { ReactElement, ReactNode } from "react";

export function renderWithUser(jsx: ReactElement) {
  return {
    user: userEvent.setup({ delay: null }),
    ...render(jsx),
  };
}

export function renderWithRouter(jsx: ReactElement) {
  const Wrapper = ({ children }: { children: ReactNode }) => (
    <RouterProvider
      router={createMemoryRouter([
        {
          path: "/",
          element: children,
        },
      ])}
    />
  );

  return {
    user: userEvent.setup({ delay: null }),
    ...render(jsx, {
      wrapper: Wrapper,
    }),
  };
}
