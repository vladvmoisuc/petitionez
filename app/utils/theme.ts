import { createTheme, colors } from "@mui/material";

const FONTS = {
  PLAYFAIR_DISPLAY: "'Playfair Display', serif",
  RED_HAT: "'Red Hat Mono', monospace",
};

export default createTheme({
  typography: {
    h1: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h2: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h3: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h4: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h5: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    fontFamily: FONTS.RED_HAT,
  },
  components: {
    MuiAppBar: {
      defaultProps: {
        color: "transparent",
        elevation: 0,
        position: "static",
      },
    },
    MuiToolbar: {
      styleOverrides: {
        root: {
          justifyContent: "space-between",
          borderBottomWidth: 1,
          borderBottomStyle: "solid",
          borderBottomColor: colors.grey[500],
        },
      },
    },
    MuiCard: {
      defaultProps: {
        elevation: 0,
        square: true,
      },
    },
    MuiSkeleton: {
      styleOverrides: {
        root: {
          backgroundColor: colors.grey[100],
        },
      },
    },
    MuiContainer: {
      defaultProps: {
        maxWidth: false,
      },
    },
  },
});
