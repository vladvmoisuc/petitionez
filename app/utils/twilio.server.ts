import sendgrid from "@sendgrid/mail";
import { readFileSync } from "fs";
import { join } from "path";

import { replace } from "~/utils/functions";

sendgrid.setApiKey(process.env.SENDGRID_API_KEY as string);

const TEMPLATES = {
  OTP: {
    from: "petitionez@gmail.com",
    subject: "Cod de autentificare petitionez.ro",
    html: readFileSync(
      join(__dirname, "..", "app", "templates", "emails", "otp.html")
    ).toString(),
  },
};

type Params = {
  type: keyof typeof TEMPLATES;
  to: string;
  data: {
    [key: string]: string;
  };
};

export default {
  send: async ({ type, to, data = {} }: Params) =>
    await sendgrid.send({
      ...TEMPLATES[type],
      to,
      html: replace(data, TEMPLATES[type].html),
    }),
};
