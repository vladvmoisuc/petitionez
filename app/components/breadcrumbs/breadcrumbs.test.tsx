import { renderWithRouter } from "~/utils/tests";

const useLocationMock = jest.fn();

jest.mock("@remix-run/react", () => ({
  useLocation: useLocationMock,
  Link: ({ children, ...props }: { children: any }) => (
    <a {...props}>{children}</a>
  ),
}));

// eslint-disable-next-line import/first
import Breadcrumbs from ".";

describe("<Breadcrumbs />", () => {
  it("should match the snapshot", () => {
    useLocationMock.mockReturnValue({
      pathname: "/campaigns",
    });

    const { container } = renderWithRouter(<Breadcrumbs />);

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when the pathname is nested", () => {
    useLocationMock.mockReturnValue({
      pathname: "/campaigns/a-cool-campaign",
    });

    const { container } = renderWithRouter(<Breadcrumbs />);

    expect(container).toMatchSnapshot();
  });
});
