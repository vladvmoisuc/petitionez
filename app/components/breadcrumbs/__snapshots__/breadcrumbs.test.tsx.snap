// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should match the snapshot 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-7a84y9-MuiTypography-root-MuiBreadcrumbs-root"
  >
    <ol
      class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <span>
          Campanii
        </span>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`<Breadcrumbs /> should match the snapshot when the pathname is nested 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-7a84y9-MuiTypography-root-MuiBreadcrumbs-root"
  >
    <ol
      class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          prefetch="intent"
          to="/campaigns"
        >
          Campanii
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <span>
          A cool campaign
        </span>
      </li>
    </ol>
  </nav>
</div>
`;
